File: crates/halo-landing/src/telemetry.rs
//! Aggregated live-telemetry source for the landing dashboard.
//!
//! Everything on `/_live/stats` and `/_live/services` ultimately resolves
//! through [`Telemetry::snapshot`]. We keep a ~1 s TTL cache in front of
//! each source so a bursty reload of the page doesn't fork a dozen
//! `rocm-smi` subprocesses.
//!
//! Sources (all 127.0.0.1 only — invariant 1/5 in `Crate-halo-landing.md`):
//! 1. `http://127.0.0.1:8180/v1/models` — loaded model name.
//! 2. `http://127.0.0.1:8180/metrics` — live tok/s, request counts.
//! 3. `rocm-smi --showtemp --showuse --json` — iGPU temp + utilisation.
//! 4. `xrt-smi examine` exit code (or `/dev/accel/accel0` presence) — NPU up.
//! 5. `~/claude output/shadow-burnin.jsonl` tail — rolling exact-match %.
//! 6. `systemctl --user is-active strix-*` — per-unit state.
//!
//! Every source degrades to a sentinel on failure rather than propagating
//! an error — this endpoint must never 5xx (invariant 5).

use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::Mutex;
use std::time::{Duration, Instant};

use serde::Serialize;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;

/// User-scope systemd units we probe for `/_live/services`. Kept small
/// and explicit — random `strix-*` units the user adds are ignored until
/// this list grows.
pub const TRACKED_SERVICES: &[&str] = &[
    "strix-server",
    "strix-landing",
    "strix-lemonade",
    "strix-echo",
    "strix-burnin",
    "strix-cloudflared",
];

/// Cached live snapshot emitted on `/_live/stats`.
///
/// `stale` flips true when the most recent `collect()` failed to reach
/// halo-server and we're serving the previous `loaded_model`.
#[derive(Debug, Clone, Serialize)]
pub struct Stats {
    pub loaded_model: String,
    pub tok_s_decode: f32,
    pub gpu_temp_c: f32,
    pub gpu_util_pct: u8,
    pub npu_up: bool,
    pub shadow_burn_exact_pct: f32,
    pub services: Vec<ServiceState>,
    pub stale: bool,
}

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct ServiceState {
    pub name: String,
    pub active: bool,
}

impl Stats {
    pub fn empty() -> Self {
        Self {
            loaded_model: String::new(),
            tok_s_decode: 0.0,
            gpu_temp_c: 0.0,
            gpu_util_pct: 0,
            npu_up: false,
            shadow_burn_exact_pct: 0.0,
            services: Vec::new(),
            stale: true,
        }
    }
}

/// Path surfaces we sniff to produce a [`Stats`]. Overridable from tests
/// so we don't need rocm-smi / halo-server running in CI.
#[derive(Clone)]
pub struct Sources {
    pub http: reqwest::Client,
    pub rocm_smi_bin: PathBuf,
    pub xrt_smi_bin: PathBuf,
    pub accel_dev: PathBuf,
    pub shadow_burnin_jsonl: PathBuf,
    pub systemctl_bin: PathBuf,
    pub services: &'static [&'static str],
    /// Base URL for halo-server probes. Tests point this at a mock.
    pub halo_server_base: String,
}

impl Default for Sources {
    fn default() -> Self {
        Self {
            http: reqwest::Client::builder()
                .user_agent("halo-landing/telemetry")
                .build()
                .expect("reqwest client"),
            rocm_smi_bin: PathBuf::from("rocm-smi"),
            xrt_smi_bin: PathBuf::from("xrt-smi"),
            accel_dev: PathBuf::from("/dev/accel/accel0"),
            shadow_burnin_jsonl: default_shadow_burnin_path(),
            systemctl_bin: PathBuf::from("systemctl"),
            services: TRACKED_SERVICES,
            halo_server_base: "http://127.0.0.1:8180".to_string(),
        }
    }
}

fn default_shadow_burnin_path() -> PathBuf {
    // `~/claude output/shadow-burnin.jsonl` — the folder name contains a
    // space on purpose (per project convention).
    let home = std::env::var("HOME").unwrap_or_else(|_| "/home/bcloud".to_string());
    PathBuf::from(home).join("claude output/shadow-burnin.jsonl")
}

/// ~1 s TTL cache around the actual source fan-out.
pub struct Telemetry {
    sources: Sources,
    cached: Mutex<Cached>,
    ttl: Duration,
}

struct Cached {
    stats: Stats,
    fetched_at: Option<Instant>,
}

impl Telemetry {
    pub fn new(sources: Sources) -> Self {
        Self {
            sources,
            cached: Mutex::new(Cached {
                stats: Stats::empty(),
                fetched_at: None,
            }),
            ttl: Duration::from_millis(1000),
        }
    }

    /// Wall-clock TTL getter — exposed for tests.
    #[cfg(test)]
    pub fn with_ttl(sources: Sources, ttl: Duration) -> Self {
        Self {
            sources,
            cached: Mutex::new(Cached {
                stats: Stats::empty(),
                fetched_at: None,
            }),
            ttl,
        }
    }

    /// Return the cached snapshot if fresh, else collect a new one.
    pub async fn snapshot(&self) -> Stats {
        {
            let c = self.cached.lock().unwrap();
            if let Some(at) = c.fetched_at
                && at.elapsed() < self.ttl
            {
                return c.stats.clone();
            }
        }

        let fresh = self.collect().await;

        let mut c = self.cached.lock().unwrap();
        c.stats = fresh.clone();
        c.fetched_at = Some(Instant::now());
        fresh
    }

    /// Bypass the cache — for tests + the SSE poll path that already
    /// throttles to 1.5 s.
    pub async fn collect(&self) -> Stats {
        let prev_model = {
            let c = self.cached.lock().unwrap();
            c.stats.loaded_model.clone()
        };

        let (model, stale) = match probe_model(&self.sources).await {
            Some(m) => (m, false),
            None => (prev_model, true),
        };
        let tok_s_decode = probe_tokps(&self.sources).await.unwrap_or(0.0);
        let (gpu_temp_c, gpu_util_pct) = probe_rocm_smi(&self.sources).await;
        let npu_up = probe_npu(&self.sources).await;
        let shadow_burn_exact_pct = probe_shadow_burn(&self.sources.shadow_burnin_jsonl).await;
        let services = probe_services(&self.sources).await;

        Stats {
            loaded_model: model,
            tok_s_decode,
            gpu_temp_c,
            gpu_util_pct,
            npu_up,
            shadow_burn_exact_pct,
            services,
            stale,
        }
    }
}

async fn probe_model(s: &Sources) -> Option<String> {
    let url = format!("{}/v1/models", s.halo_server_base);
    let r = s
        .http
        .get(&url)
        .timeout(Duration::from_millis(1500))
        .send()
        .await
        .ok()?;
    if !r.status().is_success() {
        return None;
    }
    let v: serde_json::Value = r.json().await.ok()?;
    v.get("data")
        .and_then(|d| d.get(0))
        .and_then(|m| m.get("id"))
        .and_then(|i| i.as_str())
        .map(|s| s.to_string())
}

async fn probe_tokps(s: &Sources) -> Option<f32> {
    let url = format!("{}/metrics", s.halo_server_base);
    let r = s
        .http
        .get(&url)
        .timeout(Duration::from_millis(1500))
        .send()
        .await
        .ok()?;
    if !r.status().is_success() {
        return None;
    }
    let v: serde_json::Value = r.json().await.ok()?;
    v.get("tokps_recent")
        .and_then(|x| x.as_f64())
        .map(|x| x as f32)
}

/// Run `rocm-smi --showtemp --showuse --json` and fish out edge temp +
/// GPU-use-%. On any failure (binary missing, non-zero exit, JSON shape
/// change) we return `(0.0, 0)` — the frontend renders that as "—".
///
/// rocm-smi 6.x JSON shape (per-card keyed):
/// ```json
/// { "card0": { "Temperature (Sensor edge) (C)": "48.0",
///              "GPU use (%)": "3" } }
/// ```
async fn probe_rocm_smi(s: &Sources) -> (f32, u8) {
    let out = Command::new(&s.rocm_smi_bin)
        .args(["--showtemp", "--showuse", "--json"])
        .stdin(Stdio::null())
        .stderr(Stdio::null())
        .output()
        .await;
    let Ok(out) = out else {
        return (0.0, 0);
    };
    if !out.status.success() {
        return (0.0, 0);
    }
    let Ok(v): Result<serde_json::Value, _> = serde_json::from_slice(&out.stdout) else {
        return (0.0, 0);
    };
    parse_rocm_smi_json(&v)
}

pub(crate) fn parse_rocm_smi_json(v: &serde_json::Value) -> (f32, u8) {
    let obj = match v.as_object() {
        Some(o) => o,
        None => return (0.0, 0),
    };
    // Pick the first card* key. Strix Halo has a single iGPU.
    let card = obj
        .iter()
        .find(|(k, _)| k.starts_with("card"))
        .map(|(_, v)| v);
    let Some(card) = card.and_then(|c| c.as_object()) else {
        return (0.0, 0);
    };
    let temp = card
        .iter()
        .find(|(k, _)| k.contains("Temperature") && k.contains("edge"))
        .and_then(|(_, v)| v.as_str())
        .and_then(|s| s.parse::<f32>().ok())
        .unwrap_or(0.0);
    let util = card
        .iter()
        .find(|(k, _)| k.contains("GPU use") || k.eq_ignore_ascii_case(&"GPU use (%)"))
        .and_then(|(_, v)| v.as_str())
        .and_then(|s| s.trim().trim_end_matches('%').parse::<u8>().ok())
        .unwrap_or(0);
    (temp, util)
}

/// NPU liveness: first try `xrt-smi examine` (fast, returns 0 when a
/// device is enumerated); if the binary is missing fall back to checking
/// that `/dev/accel/accel0` exists.
async fn probe_npu(s: &Sources) -> bool {
    if let Ok(out) = Command::new(&s.xrt_smi_bin)
        .arg("examine")
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .await
    {
        return out.success();
    }
    Path::new(&s.accel_dev).exists()
}

/// Tail the shadow-burnin jsonl and compute the `full_match` ratio over
/// the last N entries. N=200 matches the "~90% byte-exact" rolling-window
/// definition in CLAUDE.md.
async fn probe_shadow_burn(path: &Path) -> f32 {
    const WINDOW: usize = 200;
    let Ok(file) = tokio::fs::File::open(path).await else {
        return 0.0;
    };
    // Cheap tail: read the whole file line-by-line, keep a ring of the
    // last WINDOW lines. shadow-burnin.jsonl is append-only with ~80 B
    // per line, so even at 100k lines this is ~8 MiB — fine.
    let mut ring: Vec<String> = Vec::with_capacity(WINDOW);
    let mut reader = BufReader::new(file).lines();
    while let Ok(Some(line)) = reader.next_line().await {
        if ring.len() == WINDOW {
            ring.remove(0);
        }
        ring.push(line);
    }
    if ring.is_empty() {
        return 0.0;
    }
    let mut matches = 0u32;
    let mut total = 0u32;
    for l in &ring {
        let Ok(v): Result<serde_json::Value, _> = serde_json::from_str(l) else {
            continue;
        };
        total += 1;
        if v.get("full_match").and_then(|b| b.as_bool()).unwrap_or(false) {
            matches += 1;
        }
    }
    if total == 0 {
        0.0
    } else {
        (matches as f32) * 100.0 / (total as f32)
    }
}

async fn probe_services(s: &Sources) -> Vec<ServiceState> {
    let mut out = Vec::with_capacity(s.services.len());
    for name in s.services {
        let active = match Command::new(&s.systemctl_bin)
            .args(["--user", "is-active", name])
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .await
        {
            Ok(st) => st.success(),
            Err(_) => false,
        };
        out.push(ServiceState {
            name: (*name).to_string(),
            active,
        });
    }
    out
}

/// Return `Some(delta)` containing the service rows whose `active` flipped
/// between `prev` and `next`, or `None` if nothing changed. Pure helper
/// so the SSE `/_live/services` handler is trivial to test.
pub fn service_delta(prev: &[ServiceState], next: &[ServiceState]) -> Option<Vec<ServiceState>> {
    let mut changes = Vec::new();
    for n in next {
        let was = prev.iter().find(|p| p.name == n.name).map(|p| p.active);
        if was != Some(n.active) {
            changes.push(n.clone());
        }
    }
    if changes.is_empty() {
        None
    } else {
        Some(changes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_rocm_smi_shape() {
        let v = serde_json::json!({
            "card0": {
                "Temperature (Sensor edge) (C)": "52.0",
                "GPU use (%)": "27"
            }
        });
        let (t, u) = parse_rocm_smi_json(&v);
        assert!((t - 52.0).abs() < 0.001);
        assert_eq!(u, 27);
    }

    #[test]
    fn parse_rocm_smi_degrades_on_garbage() {
        let v = serde_json::json!({ "card0": { "bogus": 1 } });
        assert_eq!(parse_rocm_smi_json(&v), (0.0, 0));
        assert_eq!(
            parse_rocm_smi_json(&serde_json::Value::Null),
            (0.0, 0)
        );
    }

    #[test]
    fn service_delta_detects_flips() {
        let prev = vec![
            ServiceState { name: "a".into(), active: true },
            ServiceState { name: "b".into(), active: false },
        ];
        let next = vec![
            ServiceState { name: "a".into(), active: true },
            ServiceState { name: "b".into(), active: true },
        ];
        let d = service_delta(&prev, &next).expect("change detected");
        assert_eq!(d.len(), 1);
        assert_eq!(d[0].name, "b");
        assert!(d[0].active);
    }

    #[test]
    fn service_delta_is_none_when_steady() {
        let s = vec![
            ServiceState { name: "a".into(), active: true },
            ServiceState { name: "b".into(), active: true },
        ];
        assert!(service_delta(&s, &s).is_none());
    }

    #[tokio::test]
    async fn shadow_burn_counts_full_match_ratio() {
        let dir = tempdir_fallback();
        let p = dir.join("shadow-burnin.jsonl");
        let body = [
            r#"{"ts":"t","prompt_idx":0,"full_match":true}"#,
            r#"{"ts":"t","prompt_idx":1,"full_match":true}"#,
            r#"{"ts":"t","prompt_idx":2,"full_match":false}"#,
            r#"{"ts":"t","prompt_idx":3,"full_match":true}"#,
        ]
        .join("\n");
        tokio::fs::write(&p, body).await.unwrap();
        let pct = probe_shadow_burn(&p).await;
        assert!((pct - 75.0).abs() < 0.01, "expected 75.0 got {pct}");
        let _ = tokio::fs::remove_file(&p).await;
    }

    #[tokio::test]
    async fn shadow_burn_missing_file_returns_zero() {
        let p = PathBuf::from("/tmp/halo-landing-nonexistent-file-xyz.jsonl");
        let _ = tokio::fs::remove_file(&p).await;
        assert_eq!(probe_shadow_burn(&p).await, 0.0);
    }

    #[tokio::test]
    async fn collect_returns_stats_with_missing_sources() {
        // Point every external at a broken/nonexistent path — collect()
        // must not panic and must yield a Stats with sentinels.
        let sources = Sources {
            http: reqwest::Client::new(),
            rocm_smi_bin: PathBuf::from("/nonexistent/rocm-smi"),
            xrt_smi_bin: PathBuf::from("/nonexistent/xrt-smi"),
            accel_dev: PathBuf::from("/nonexistent/accel0"),
            shadow_burnin_jsonl: PathBuf::from("/nonexistent/shadow.jsonl"),
            systemctl_bin: PathBuf::from("/nonexistent/systemctl"),
            services: TRACKED_SERVICES,
            // Unroutable TEST-NET-1 address so probe_model fails fast
            // without actually hitting halo-server on the dev box.
            halo_server_base: "http://192.0.2.1:1".to_string(),
        };
        let t = Telemetry::new(sources);
        let s = t.collect().await;
        // loaded_model blank, stale flagged (halo-server unreachable).
        assert!(s.loaded_model.is_empty());
        assert!(s.stale);
        assert_eq!(s.tok_s_decode, 0.0);
        assert_eq!(s.gpu_temp_c, 0.0);
        assert_eq!(s.gpu_util_pct, 0);
        assert!(!s.npu_up);
        assert_eq!(s.shadow_burn_exact_pct, 0.0);
        // Services must include every tracked entry, all false.
        assert_eq!(s.services.len(), TRACKED_SERVICES.len());
        assert!(s.services.iter().all(|svc| !svc.active));
    }

    #[tokio::test]
    async fn snapshot_respects_cache_ttl() {
        let sources = Sources {
            http: reqwest::Client::new(),
            rocm_smi_bin: PathBuf::from("/nonexistent/rocm-smi"),
            xrt_smi_bin: PathBuf::from("/nonexistent/xrt-smi"),
            accel_dev: PathBuf::from("/nonexistent/accel0"),
            shadow_burnin_jsonl: PathBuf::from("/nonexistent/shadow.jsonl"),
            systemctl_bin: PathBuf::from("/nonexistent/systemctl"),
            services: TRACKED_SERVICES,
            halo_server_base: "http://192.0.2.1:1".to_string(),
        };
        let t = Telemetry::with_ttl(sources, Duration::from_secs(60));
        let a = t.snapshot().await;
        let b = t.snapshot().await;
        // Same stale+empty data, re-used from cache.
        assert_eq!(a.loaded_model, b.loaded_model);
        assert_eq!(a.stale, b.stale);
    }

    fn tempdir_fallback() -> PathBuf {
        let p = PathBuf::from(format!(
            "/tmp/halo-landing-test-{}",
            std::process::id()
        ));
        std::fs::create_dir_all(&p).unwrap();
        p
    }
}

File: cpp/landing/src/telemetry.cpp
#include "onebit/landing/telemetry.hpp"

#include <httplib.h>

#include <algorithm>
#include <array>
#include <cctype>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <deque>
#include <fstream>
#include <regex>
#include <sstream>
#include <string>

namespace onebit::landing {

namespace {

using json = nlohmann::json;

// ---- subprocess helper -----------------------------------------------------
// popen() captures stdout into a string and returns the exit status.
// Stderr is silenced (`2>/dev/null`) to match the Rust port's
// `Stdio::null()` pipes.
struct ExecResult {
    int         status{-1};
    std::string stdout_;
};

[[nodiscard]] ExecResult exec_capture(const std::string& cmd)
{
    ExecResult r;
    std::FILE* f = popen(cmd.c_str(), "r");
    if (!f) {
        return r;
    }
    char buf[4096];
    while (auto n = std::fread(buf, 1, sizeof(buf), f)) {
        r.stdout_.append(buf, n);
    }
    int rc = pclose(f);
    if (rc == -1) {
        r.status = -1;
    } else {
        // WEXITSTATUS portable enough on Linux.
        r.status = (rc >> 8) & 0xff;
    }
    return r;
}

// Returns true if `path` exists at the filesystem level (no perms check).
[[nodiscard]] bool path_exists(const std::filesystem::path& p) noexcept
{
    std::error_code ec;
    return std::filesystem::exists(p, ec);
}

[[nodiscard]] std::string shell_quote(const std::string& s)
{
    std::string out{"'"};
    for (char c : s) {
        if (c == '\'') {
            out += "'\\''";
        } else {
            out += c;
        }
    }
    out += "'";
    return out;
}

// Split an http(s) URL into host+port for httplib::Client.
struct UrlParts {
    std::string scheme;
    std::string host;
    int         port{-1};
    bool        ok{false};
};
[[nodiscard]] UrlParts split_url(std::string_view url)
{
    UrlParts u;
    static const std::regex re(R"(^(https?)://([^:/]+)(?::(\d+))?(?:/.*)?$)");
    std::cmatch m;
    if (!std::regex_match(url.data(), url.data() + url.size(), m, re)) {
        return u;
    }
    u.scheme = m[1].str();
    u.host   = m[2].str();
    u.port   = m[3].matched ? std::stoi(m[3].str())
                            : (u.scheme == "https" ? 443 : 80);
    u.ok = true;
    return u;
}

[[nodiscard]] std::optional<json> http_get_json(const std::string& base,
                                                std::string_view   path,
                                                std::chrono::milliseconds timeout)
{
    auto u = split_url(base);
    if (!u.ok) {
        return std::nullopt;
    }
    httplib::Client cli(u.scheme + "://" + u.host + ":" + std::to_string(u.port));
    cli.set_connection_timeout(timeout);
    cli.set_read_timeout(timeout);
    cli.set_write_timeout(timeout);
    auto res = cli.Get(std::string{path});
    if (!res || res->status < 200 || res->status >= 300) {
        return std::nullopt;
    }
    try {
        return json::parse(res->body);
    } catch (const json::parse_error&) {
        return std::nullopt;
    }
}

} // namespace

// ---- Sources ---------------------------------------------------------------

Sources Sources::defaults()
{
    Sources s;
    // ~/claude output/shadow-burnin.jsonl (folder name has a literal space —
    // project convention from CLAUDE.md "Benchmark output folder").
    const char* home = std::getenv("HOME");
    std::filesystem::path home_p = home ? home : "/home/bcloud";
    s.shadow_burnin_jsonl = home_p / "claude output" / "shadow-burnin.jsonl";
    s.services.reserve(TRACKED_SERVICES.size());
    for (const auto& sv : TRACKED_SERVICES) {
        s.services.emplace_back(sv);
    }
    return s;
}

// ---- service_delta ---------------------------------------------------------

std::optional<std::vector<ServiceState>>
service_delta(const std::vector<ServiceState>& prev,
              const std::vector<ServiceState>& next)
{
    std::vector<ServiceState> changes;
    for (const auto& n : next) {
        auto it = std::find_if(prev.begin(), prev.end(),
                               [&](const ServiceState& p) { return p.name == n.name; });
        const bool was_known    = (it != prev.end());
        const bool prev_active  = was_known ? it->active : !n.active; // sentinel: anything != n.active
        if (!was_known || prev_active != n.active) {
            changes.push_back(n);
        }
    }
    if (changes.empty()) {
        return std::nullopt;
    }
    return changes;
}

// ---- parse_rocm_smi_json ---------------------------------------------------

std::pair<float, std::uint8_t> parse_rocm_smi_json(const json& v)
{
    if (!v.is_object()) {
        return {0.0F, 0};
    }
    // Find the first "card*" key.
    const json* card = nullptr;
    for (auto it = v.begin(); it != v.end(); ++it) {
        const auto& key = it.key();
        if (key.rfind("card", 0) == 0) {
            card = &it.value();
            break;
        }
    }
    if (!card || !card->is_object()) {
        return {0.0F, 0};
    }

    float        temp = 0.0F;
    std::uint8_t util = 0;

    for (auto it = card->begin(); it != card->end(); ++it) {
        const std::string& k = it.key();
        if (!it.value().is_string()) {
            continue;
        }
        const std::string vstr = it.value().get<std::string>();
        if (k.find("Temperature") != std::string::npos
            && k.find("edge") != std::string::npos)
        {
            try { temp = std::stof(vstr); }
            catch (...) { temp = 0.0F; }
        } else if (k.find("GPU use") != std::string::npos) {
            // Trim whitespace + trailing '%' then parse.
            std::string trimmed = vstr;
            while (!trimmed.empty() && std::isspace(static_cast<unsigned char>(trimmed.front()))) {
                trimmed.erase(trimmed.begin());
            }
            while (!trimmed.empty() && (std::isspace(static_cast<unsigned char>(trimmed.back()))
                                         || trimmed.back() == '%'))
            {
                trimmed.pop_back();
            }
            try {
                int n = std::stoi(trimmed);
                if (n < 0)        { util = 0; }
                else if (n > 255) { util = 255; }
                else              { util = static_cast<std::uint8_t>(n); }
            } catch (...) { util = 0; }
        }
    }
    return {temp, util};
}

// ---- HTTP probes -----------------------------------------------------------

std::optional<std::string> probe_model(const Sources& s)
{
    auto v = http_get_json(s.onebit_server_base, "/v1/models",
                           std::chrono::milliseconds{1500});
    if (!v) { return std::nullopt; }
    auto data = v->find("data");
    if (data == v->end() || !data->is_array() || data->empty()) {
        return std::nullopt;
    }
    const auto& first = (*data)[0];
    auto id = first.find("id");
    if (id == first.end() || !id->is_string()) {
        return std::nullopt;
    }
    return id->get<std::string>();
}

std::optional<float> probe_tokps(const Sources& s)
{
    auto v = http_get_json(s.onebit_server_base, "/metrics",
                           std::chrono::milliseconds{1500});
    if (!v) { return std::nullopt; }
    auto t = v->find("tokps_recent");
    if (t == v->end() || !t->is_number()) { return std::nullopt; }
    return static_cast<float>(t->get<double>());
}

// ---- subprocess + filesystem probes ---------------------------------------

std::pair<float, std::uint8_t> probe_rocm_smi(const Sources& s)
{
    if (!std::filesystem::exists(s.rocm_smi_bin)
        && s.rocm_smi_bin.is_absolute())
    {
        return {0.0F, 0};
    }
    const std::string cmd =
        shell_quote(s.rocm_smi_bin.string())
        + " --showtemp --showuse --json 2>/dev/null </dev/null";
    auto r = exec_capture(cmd);
    if (r.status != 0 || r.stdout_.empty()) {
        return {0.0F, 0};
    }
    try {
        auto v = json::parse(r.stdout_);
        return parse_rocm_smi_json(v);
    } catch (const json::parse_error&) {
        return {0.0F, 0};
    }
}

bool probe_npu(const Sources& s)
{
    if (s.xrt_smi_bin.is_absolute() && !path_exists(s.xrt_smi_bin)) {
        return path_exists(s.accel_dev);
    }
    const std::string cmd =
        shell_quote(s.xrt_smi_bin.string())
        + " examine >/dev/null 2>/dev/null </dev/null";
    auto r = exec_capture(cmd);
    if (r.status == 0) {
        return true;
    }
    // Fallback to /dev/accel/accel0 presence (matches Rust path).
    return path_exists(s.accel_dev);
}

float probe_shadow_burn(const std::filesystem::path& p)
{
    constexpr std::size_t WINDOW = 200;
    std::ifstream in(p);
    if (!in.is_open()) {
        return 0.0F;
    }
    std::deque<std::string> ring;
    std::string line;
    while (std::getline(in, line)) {
        if (ring.size() == WINDOW) {
            ring.pop_front();
        }
        ring.push_back(line);
    }
    if (ring.empty()) {
        return 0.0F;
    }
    std::uint32_t matches = 0;
    std::uint32_t total   = 0;
    for (const auto& l : ring) {
        if (l.empty()) { continue; }
        try {
            auto v = json::parse(l);
            ++total;
            auto it = v.find("full_match");
            if (it != v.end() && it->is_boolean() && it->get<bool>()) {
                ++matches;
            }
        } catch (const json::parse_error&) {
            // skip — matches Rust's `continue`.
        }
    }
    if (total == 0) {
        return 0.0F;
    }
    return static_cast<float>(matches) * 100.0F / static_cast<float>(total);
}

std::vector<ServiceState> probe_services(const Sources& s)
{
    std::vector<ServiceState> out;
    out.reserve(s.services.size());
    const bool sysctl_missing =
        s.systemctl_bin.is_absolute() && !path_exists(s.systemctl_bin);
    for (const auto& name : s.services) {
        bool active = false;
        if (!sysctl_missing) {
            const std::string cmd =
                shell_quote(s.systemctl_bin.string())
                + " --user is-active " + shell_quote(name)
                + " >/dev/null 2>/dev/null </dev/null";
            auto r = exec_capture(cmd);
            active = (r.status == 0);
        }
        out.push_back(ServiceState{name, active});
    }
    return out;
}

// ---- Telemetry -------------------------------------------------------------

Telemetry::Telemetry(Sources sources, std::chrono::milliseconds ttl)
    : sources_(std::move(sources)), ttl_(ttl) {}

Stats Telemetry::snapshot()
{
    {
        std::lock_guard lk(mu_);
        if (fetched_at_.has_value()) {
            const auto age = std::chrono::steady_clock::now() - *fetched_at_;
            if (age < ttl_) {
                return cached_;
            }
        }
    }

    Stats fresh = collect();

    std::lock_guard lk(mu_);
    cached_     = fresh;
    fetched_at_ = std::chrono::steady_clock::now();
    return fresh;
}

Stats Telemetry::collect()
{
    std::string prev_model;
    {
        std::lock_guard lk(mu_);
        prev_model = cached_.loaded_model;
    }

    auto model_opt = probe_model(sources_);
    std::string model = model_opt.value_or(prev_model);
    bool stale        = !model_opt.has_value();

    float tok_s_decode  = probe_tokps(sources_).value_or(0.0F);
    auto [temp, util]   = probe_rocm_smi(sources_);
    bool npu_up         = probe_npu(sources_);
    float burn          = probe_shadow_burn(sources_.shadow_burnin_jsonl);
    auto services       = probe_services(sources_);

    Stats s;
    s.loaded_model          = std::move(model);
    s.tok_s_decode          = tok_s_decode;
    s.gpu_temp_c            = temp;
    s.gpu_util_pct          = util;
    s.npu_up                = npu_up;
    s.shadow_burn_exact_pct = burn;
    s.services              = std::move(services);
    s.stale                 = stale;
    return s;
}

} // namespace onebit::landing
